plugins/reachability/src/main/java/org/qbicc/plugin/reachability/ReachabilityBlockBuilder.java

```diff
@@ -28,135 +36,196 @@
+        @Override
+        public Void visitUnknown(ReachabilityContext param, Action node) {
+            visitUnknown(param, (Node) node);
+            return null;
+        }
 
-    @Override
-    public BasicBlock callNoReturn(ValueHandle target, List<Value> arguments) {
-        target.accept(this, null);
-        return super.callNoReturn(target, arguments);
-    }
+        @Override
+        public Void visitUnknown(ReachabilityContext param, Value node) {
+            visitUnknown(param, (Node) node);
+            return null;
+        }
 
-    @Override
-    public BasicBlock invokeNoReturn(ValueHandle target, List<Value> arguments, BlockLabel catchLabel) {
-        target.accept(this, null);
-        return super.invokeNoReturn(target, arguments, catchLabel);
-    }
+        @Override
+        public Void visitUnknown(ReachabilityContext param, ValueHandle node) {
+            visitUnknown(param, (Node) node);
+            return null;
+        }
 
-    @Override
-    public BasicBlock tailCall(ValueHandle target, List<Value> arguments) {
-        target.accept(this, null);
-        return super.tailCall(target, arguments);
-    }
+        @Override
+        public Void visitUnknown(ReachabilityContext param, Terminator node) {
+            if (visitUnknown(param, (Node) node)) {
+                // process reachable successors
+                int cnt = node.getSuccessorCount();
+                for (int i = 0; i < cnt; i ++) {
+                    node.getSuccessor(i).getTerminator().accept(this, param);
+                }
+            }
+            return null;
+        }
 
-    @Override
-    public BasicBlock tailInvoke(ValueHandle target, List<Value> arguments, BlockLabel catchLabel) {
-        target.accept(this, null);
-        return super.tailInvoke(target, arguments, catchLabel);
-    }
+        boolean visitUnknown(ReachabilityContext param, Node node) {
+            if (param.visited.add(node)) {
+                if (node.hasValueHandleDependency()) {
+                    node.getValueHandle().accept(this, param);
+                }
+                int cnt = node.getValueDependencyCount();
+                for (int i = 0; i < cnt; i ++) {
+                    node.getValueDependency(i).accept(this, param);
+                }
+                if (node instanceof OrderedNode) {
+                    Node dependency = ((OrderedNode) node).getDependency();
+                    if (dependency instanceof Action) {
+                        ((Action) dependency).accept(this, param);
+                    } else if (dependency instanceof Value) {
+                        ((Value) dependency).accept(this, param);
+                    } else if (dependency instanceof Terminator) {
+                        ((Terminator) dependency).accept(this, param);
+                    } else if (dependency instanceof ValueHandle) {
+                        ((ValueHandle) dependency).accept(this, param);
+                    }
```

Comment:
If you do go ahead and remove the `Invoke.ReturnValue` bit below, you might as well update this to be Java 17-nice.  Otherwise it's OK as is if you don't want to do that.